### PR TITLE
Restructure -Wno switches, silence bogus warnings in external headers

### PR DESCRIPTION
**Pull request by tamiko**

### CODE REVIEW

File: cmake/configure/configure_2_trilinos.cmake [hunk 1]
```diff
@@ -192,13 +192,14 @@ MACRO(FEATURE_TRILINOS_CONFIGURE_EXTERNAL)
   SET(DEAL_II_EXPAND_TRILINOS_MPI_BLOCKVECTOR "TrilinosWrappers::MPI::BlockVector")
   SET(DEAL_II_EXPAND_TRILINOS_MPI_VECTOR "TrilinosWrappers::MPI::Vector")
 
+  ENABLE_IF_SUPPORTED(TRILINOS_CXX_FLAGS "-Wno-extra")
```
Comment (bangerth) [inline cmake/configure/configure_2_trilinos.cmake]: So you want this flag to be set regardless of the Trilinos version?

File: cmake/configure/configure_boost.cmake [hunk 2]
```diff
@@ -22,8 +22,29 @@ SET(DEAL_II_WITH_BOOST ON # Always true. We need it :-]
...
+  # compiler supports it.
+  #
+  ENABLE_IF_SUPPORTED(BOOST_CXX_FLAGS "-Wno-unused-local-typedefs")
+
+  IF(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    # not harmless but needed for boost <1.50.0
+    ENABLE_IF_SUPPORTED(BOOST_CXX_FLAGS "-Wno-delete-non-virtual-dtor")
+
+    # suppress warnings in boost 1.56:
+    ENABLE_IF_SUPPORTED(BOOST_CXX_FLAGS "-Wno-c++11-extensions")
```
Comment (bangerth) [inline cmake/configure/configure_boost.cmake]: What happens with this flag if we enable C++11?

File: cmake/configure/configure_2_trilinos.cmake (hunk 1 above)
Comment (tamiko) [inline cmake/configure/configure_2_trilinos.cmake]: > So you want this flag to be set regardless of the Trilinos version?

Yes. For the time being.

File: cmake/configure/configure_boost.cmake (hunk 2 above)
Comment (tamiko) [inline cmake/configure/configure_boost.cmake]: There is no warning "[...] is a C++11 extension" in C++11 mode. It is omitted, I guess.

Comment (tamiko) [inline cmake/configure/configure_boost.cmake]: @bangerth: I'll have a look at the clang specific <code>-Wno</code> switches sometime later. Maybe some of them can be removed.

File: cmake/configure/configure_2_trilinos.cmake (hunk 1 above)
Comment (tamiko) [inline cmake/configure/configure_2_trilinos.cmake]: Let the time being, be 30 minutes.

**Comment by tamiko** [comment]
@bangerth updated.

File: cmake/configure/configure_2_trilinos.cmake (hunk 1 above)
Comment (bangerth) [inline cmake/configure/configure_2_trilinos.cmake]: On 04/15/2015 11:33 AM, Matthias Maier wrote:
> So you want this flag to be set regardless of the Trilinos version?
> Yes. For the time being.

No, I don't care. I was just wondering whether that's what you intended.

File: cmake/setup_compiler_flags_gnu.cmake [hunk 3]
```diff
@@ -100,19 +92,12 @@ IF(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
...
   ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-Wno-unsupported-friend")
   ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-Wno-unused-function")
-  ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-Wno-unused-parameter") # mainly caused by BOOST
   ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-Wno-unused-private-field")
   ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-Wno-unused-variable")
-
-  # suppress warnings in boost 1.56:
-  ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-Wno-c++11-extensions")
-  ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-Wno-c99-extensions")
-  ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-Wno-variadic-macros")
```
Comment (bangerth) [inline cmake/setup_compiler_flags_gnu.cmake]: So the problem I see with distributing these `ENABLE_IF_SUPPORTED` checks to the configuration of the sub-libraries is (i) that it's going to much harder to find out where exactly a flag is set, and (ii) that you need to adjust the configuration of sub-libraries or external dependencies for the various compilers we support -- i.e., that the BOOST configuration suddenly has to have switches in it for GCC, CLANG, etc. I find that not overly elegant.

I don't know how to best solve this. Your approach may well be the best one can do. I just wanted to raise the issue.

Comment (tamiko) [inline cmake/setup_compiler_flags_gnu.cmake]: At least it is now under <code>BOOST_CXX_FLAGS</code> in <code>detailed.log</code>.

**Comment by tamiko** [comment]
@bangerth What about we discuss any restructuring in August?

I would like to see this merged soon as it reduces the current number of warnings from roughly 900 to roughly 50 (or something on these lines).

**Comment by bangerth** [comment]
I'm not opposed to your approach. If you have thought through the issues, then please go ahead and merge. I just wanted to share an observation :-)

**Comment by tamiko** [comment]
Thanks. :-)

**Pull request merged**